overlord/servicestate/quota_control.go

```diff
@@ -0,0 +1,260 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package servicestate
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/gadget/quantity"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/progress"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/quota"
+	"github.com/snapcore/snapd/snapdenv"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/wrappers"
+)
+
+func ensureSnapServicesForGroup(st *state.State, grp *quota.Group, allGrps map[string]*quota.Group) error {
+	// build the map of snap infos to options to provide to EnsureSnapServices
+	snapSvcMap := map[*snap.Info]*wrappers.SnapServiceOptions{}
+	for _, sn := range grp.Snaps {
+		info, err := snapstate.CurrentInfo(st, sn)
+		if err != nil {
+			return err
+		}
+
+		opts, err := SnapServiceOptions(st, sn, allGrps)
+		if err != nil {
+			return err
+		}
+
+		snapSvcMap[info] = opts
+	}
+
+	// TODO: the following lines should maybe be EnsureOptionsForDevice() or
+	// something since it is duplicated a few places
+	ensureOpts := &wrappers.EnsureSnapServicesOptions{
+		Preseeding: snapdenv.Preseeding(),
+	}
+
+	// set RequireMountedSnapdSnap if we are on UC18+ only
+	deviceCtx, err := snapstate.DeviceCtx(st, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	if !deviceCtx.Classic() && deviceCtx.Model().Base() != "" {
+		ensureOpts.RequireMountedSnapdSnap = true
+	}
+
+	// TODO: do we need to restart modified services ?
+	return wrappers.EnsureSnapServices(snapSvcMap, ensureOpts, nil, progress.Null)
+}
+
+func validateSnapForAddingToGroup(st *state.State, name string, allGrps map[string]*quota.Group) error {
```

Comment:
This helper gets called for every snap separately, we might consider to improve this later in a followup with a version that takes all snaps and builds a simple map lookup first, then goes over them once.

Reply:
done in https://github.com/snapcore/snapd/pull/10216